### ISSUE TITLE
<Description of the proposed change>

### ISSUE THREAD

**Issue by lapp0** [outlines-dev/outlines#678]
### What behavior of the library made you think about the improvement?

Outlines supports 5 inference engines (Ignoring OpenAI / Azure). We should ensure their interfaces and behaviors are identical. We are accumulating technical debt because of their inconsistencies.

For example, I'm hesitant to implement https://github.com/outlines-dev/outlines/issues/644 / https://github.com/outlines-dev/outlines/issues/647 because I would have to implement it for three distinct interfaces. Another example is the fact that https://github.com/outlines-dev/outlines/issues/614 doesn't work with vLLM or LlamaCpp.

A breakdown of the discrepancies follows.

Logit modification:
- For Transformers / ExLlamaV2 / Mamba logits are modified via an FSM passed to `SequenceGenerator`.
- For LlamaCpp logits are modified via logits processors defined in `outlines.models.llamacpp`.
- For vLLM logits are modified via logits processors defined in `outlines.serve.vllm`. vLLMs logits processors repeat substantial portions of LlamaCpps logits processors.

Sequence Generation:
- For Transformers / ExLlamaV2 / Mamba `model()` generates the logits, and SequenceGenerator samples and manages the generation.
- For LlamaCpp, the model is functionally the `SequenceGenerator`, and `outlines.generate.regex_llamacpp` patches the model.
- There is no `outlines.models.vllm`. Outlines provides a web server and the logits processors. `vllm` does all the work.

Tokenizer:
- Transformers / ExLlamaV2 / Mamba use a TransformersTokenizer
- LlamaCpp uses a custom tokenizer compatible with `outlines.models.tokenizer.Tokenizer`
- vLLM patches the tokenizer within the logits processor to be compatible

### How would you like it to behave?

Unification should look as follows:

#### Step 1: https://github.com/outlines-dev/outlines/pull/676
- Ensure LlamaCpp and vLLM use the same logits processors
- Ensure LlamaCpp and vLLM both use a tokenizer subclassing `outlines.models.tokenizer.Tokenizer`
  - don't monkeypatch vLLM tokenizer in the logits processor itself

Result: logits processors have a single canonical implementation

#### Step 2:
- Ensure `sequence_generator` applies `LogitsProcessors`, and doesn't manage FSMs
  - The appropriate logits processor is passed based on the `outlines.generate` function

Result: all 5 models use logits processors, the business logic for constraining the next token based on an FSM is encapsulated in one place

#### Step 3:
- Implement `outlines.models.vllm`
  - `outlines.models.vllm.__call__` returns logits and kv cache, and can be used with `SequenceProcessor`
- `outlines.models.llamacpp.__call__` returns logits and kv cache, and can be used with `SequenceProcessor`

Step 3b:
- Remove `outlines.generate.regex_llamacpp` (same for other `outlines.generate.*_llamacpp`), interface should work with `outlines.generate.regex`
  - Involves subclassing `llama_cpp.Llama` such that `llama.Llama.generate()` doesn't call `Llama.sample()` but instead allows Outlines to handle the sampling logic.

Result: All 5 models compatible with `SequenceGenerator`, model agnostic `outlines.generate`, `outlines.models.vllm` available

#### Step 4:
- "Serve Any Model": https://github.com/outlines-dev/outlines/issues/655

Result: Any model servable via `outlines.serve`

**Comment by rlouf** [comment]
> Additionally, `outlines.generate.regex_llamacpp` exists to enforce a sampler constraint on `llamacpp`, however we have a bug resulting from the fact that `outlines.generate.json_llamacpp` doesn't exist and doesn't enforce the same sampler constraint.

What bug? `outlines.generate.json` is the default function so it should work for both `Transformers` and `LlamaCpp`.

> * For Transformers / ExLlamaV2 / Mamba `model()` generates the logits, and `SequenceGenerator` samples and manages the generation.
> * For LlamaCpp, the model is functionally the `SequenceGenerator`, and `outlines.generate.regex_llamacpp` patches the model to apply logits processors. `SequenceGenerator` is not used.

The difference between llama.cpp and transformers was temporary; it was quicker to integrate via logits processors than fixing the existing implementation. Eventually it will use `SequenceGenerator`.

> * There is no `outlines.models.vllm`. Outlines provides a web server and the logits processors. `vllm` does all the work.

That could be easily added. However all model calls in Outlines are synchronous for now, so we wouldn't benefit from vLLM's continuous batching feature.


> * Ensure LlamaCpp and vLLM use the same logits processors
> * Ensure LlamaCpp and vLLM both use a tokenizer subclassing `outlines.models.tokenizer.Tokenizer`

Aren't you moving complexity from one part of the codebase to another? As long as `transformers`, `llama-cpp-python` and `vLLM` use different interfaces for their tokenizer there will be some special casing.

> Result: all 5 models use logits processors, the business logic for constraining the next token based on an FSM is encapsulated in one place

I am not opposed to using logits processors in `sequence_generator` instead of having it manage the FSM. However I doubt this will solve anything? The OpenAI API encapsulates the text generation process, so do vLLM and llama.cpp (for now). You will always need to dispatch depending on the model at the generation level. 

We can always "clean" the `generate` module by dispatching in the way [aesara](https://github.com/aesara-devs/aesara) does dispatch Ops for different backends. That is the dispatching logic is implemented where the model is instead of closer to the generation. It comes with trade-offs, but I think it would indeed be cleaner.


>   * `outlines.models.vllm.__call__` returns logits and kv cache, and can be used with `SequenceGenerator`
> * `outlines.models.llamacpp.__call__` returns logits and kv cache, and can be used with `SequenceGenerator`

Ok if there's an easy way to do this.

> * Remove `outlines.generate.regex_llamacpp` (same for other `outlines.generate.*_llamacpp`), interface should work with `outlines.generate.regex`

It does as far as the user is concerned.

>   * Involves subclassing `llama_cpp.Llama` such that `llama.Llama.generate()` doesn't call `Llama.sample()` but instead allows Outlines to handle the sampling logic.

That's planned. Again, using `sample` was a temporary fix for problems that many users were having. There are subtleties linked to KV-cache management when it comes to integrating `llama-cpp-python` and that will require some work.

> Result: All 5 models compatible with `SequenceGenerator`, model agnostic `outlines.generate`, `outlines.models.vllm` available

How does that work with OpenAI?


> * Inference engine integration tests: every inference engine X every sampler (related: [beam search doesn't work with ExLlamaV2)](https://github.com/outlines-dev/outlines/pull/676#issuecomment-1953584549)

I'm fine with that, but we'll need to do something to handle the ever-growing test suite duration.
